compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -59,18 +61,28 @@ class ExtractSemanticDB private (phaseMode: ExtractSemanticDB.PhaseMode, suffix:
   // Check not needed since it does not transform trees
   override def isCheckable: Boolean = false
 
-  override def run(using Context): Unit =
-    val unit = ctx.compilationUnit
-    if (phaseMode == ExtractSemanticDB.PhaseMode.PostTyper)
-      val extractor = ExtractSemanticDB.Extractor()
-      extractor.extract(unit.tpdTree)
-      ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
-    else
-      val warnings = ctx.reporter.allWarnings.collect {
-        case w if w.pos.source == ctx.source => w.toSemanticDiagnostic
-      }
-      if (warnings.nonEmpty)
-        ExtractSemanticDB.appendDiagnostics(unit.source, warnings)
+  override def runOn(units: List[CompilationUnit])(using ctx: Context): List[CompilationUnit] = {
+    val appendWarnings = phaseMode == ExtractSemanticDB.PhaseMode.PostInlining
+    val warnings =
+      if (appendWarnings)
+        ctx.reporter.allWarnings.groupBy(w => w.pos.source)
+      else Map.empty
+
+    units.asJava.parallelStream().map { unit =>
```

Comment:
unfortunately it isn't safe to parallelise the semanticDB extractor. What I suggest is to extract and convert the warnings to SemanticDiagnostic linearly, and then the `ExtractSemanticDB.appendDiagnostics` should be safe to do in parallel (please ensure it doesnt use any compiler internals, just read/write semanticdb)